common/metadata-helper.cpp

```diff
@@ -0,0 +1,324 @@
+            {
+                guid_to_devices[d.guid].push_back(d);
+            }
+
+            for (auto&& kvp : guid_to_devices)
+            {
+                auto guid = kvp.first;
+
+                std::stringstream ss;
+                ss << "SYSTEM\\CurrentControlSet\\Control\\DeviceClasses\\{" << guid << "}";
+                auto s = ss.str();
+                std::wstring ws(s.begin(), s.end());
+
+                HKEY hKey;
+                if (RegOpenKeyEx(HKEY_LOCAL_MACHINE, ws.c_str(), 0, KEY_READ | KEY_WOW64_64KEY, &hKey) == ERROR_SUCCESS)
+                {
+                    // Don't forget to release in the end:
+                    std::shared_ptr<void> raii(hKey, RegCloseKey);
+
+                    TCHAR    achClass[MAX_PATH] = TEXT("");  // buffer for class name 
+                    DWORD    cchClassName = MAX_PATH;  // size of class string 
+                    DWORD    cSubKeys = 0;               // number of subkeys 
+                    DWORD    cbMaxSubKey;              // longest subkey size 
+                    DWORD    cchMaxClass;              // longest class string 
+                    DWORD    cValues;              // number of values for key 
+                    DWORD    cchMaxValue;          // longest value name 
+                    DWORD    cbMaxValueData;       // longest value data 
+                    DWORD    cbSecurityDescriptor; // size of security descriptor 
+                    FILETIME ftLastWriteTime;      // last write time 
+
+                    DWORD retCode = RegQueryInfoKey(
+                        hKey,                    // key handle 
+                        achClass,                // buffer for class name 
+                        &cchClassName,           // size of class string 
+                        NULL,                    // reserved 
+                        &cSubKeys,               // number of subkeys 
+                        &cbMaxSubKey,            // longest subkey size 
+                        &cchMaxClass,            // longest class string 
+                        &cValues,                // number of values for this key 
+                        &cchMaxValue,            // longest value name 
+                        &cbMaxValueData,         // longest value data 
+                        &cbSecurityDescriptor,   // security descriptor 
+                        &ftLastWriteTime);       // last write time 
+
+                    for (int i = 0; i<cSubKeys; i++)
+                    {
+                        TCHAR achKey[MAX_KEY_LENGTH];
+                        DWORD cbName = MAX_KEY_LENGTH;
+                        retCode = RegEnumKeyEx(hKey, i,
+                            achKey,
+                            &cbName,
+                            NULL,
+                            NULL,
+                            NULL,
+                            &ftLastWriteTime);
+                        if (retCode == ERROR_SUCCESS)
```

Comment:
In case of errors, do we usually LOG_DEBUG() it?
What about the successes? Do we want to debug-log any of this so we can see what happened?